# Data Wrangling

Here are some fun facts about your Python skills after 1.5 weeks:

+ You can create and manipulate a Jupyter notebook.
+ You can download a CSV file to your computer and upload it to JupyterHub.
+ You can read a CSV file to create a `DataFrame`.
+ You can call `head` to get a new shorter `DataFrame` with just the top few rows.
+ You can extract a column from a `DataFrame`, and you're aware that the column type is `Series`.
+ You can, given an example, figure out how use a Python dictionary to create a `DataFrame`.
+ You can create a list of values in Python and name it using an assignment statement.
+ You can calculate the average of a `Series` or a list, and can find the maximum and minimum values using built-in functions.
+ You can name any data you want using variables.
+ You can print any data you have.

# What's next?

__Subsetting__:

+ Create a new `DataFrame` containing a subset of the columns.
+ Create a new `DataFrame` containing a subset of the rows
  - by index, using slicing
  - by looking for rows that match our criteria
+ Renaming columns

# GGR: Time Use dataset

Dataset: Statistics Canada General Social Survey's (GSS) Time Use (TU) Survey

+ Tracks how people spend their time
+ two parts
  - an 'episode' file where each row describes an event for one person. Each person has one row per event (like "make coffee") that they experienced.
  - a 'main' file that includes meta-information about the individuals and also includes summary information from the 'episode' file. There is 1 row of information per person.
+ we will stick to the 'main' file for now
The data are stored in file `gss_tu2016_main_file.csv`. The name uses abbreviations:
- GSS: general social survey
- tu2016: time use from the year 2016
- csv: comma-separated values

## 'main' file

- Summarizes time use information
- Includes summary information from the 'episode' file
- Includes information from other survey questions about the participant's socioeconomic status (demographics, economic situation, self-reported health status, etc.).
- Each row describes a single participant using information taken from the survey or summarized from the 'episode' file.

| ID  | Home Province | Self-rated Health | Time Spent Sleeping | Time Spent Commuting | Time Spent Eating | ... | Count of Sleeping Activities | Count of Commuting Activities | Count of Eating Activities |
|-----|---------------|-------------------|---------------------|----------------------|-------------------|-----|------------------------------|-------------------------------|----------------------------|
| 1   | ON            | Excellent         | 440 min             | 60 min               | 70 min            |     | 2                            | 2                             | 3                          |
| 2   | AB            | Poor              | 600 min             | 80 min               | 130 min           |     | 3                            | 4                             | 2                          |
| ... |               |                   |                     |                      |                   |     |                              |                               |                            |
| 238 | BC            | Fair              | 380 min             | 20 min               | 40 min            |     | 2                            | 2                             | 4                          |

The 'main' data are stored in file `gss_tu2016_main_file.csv`. It is in the same folder as this notebook.

The name uses abbreviations:
- GSS: general social survey
- tu2016: time use from the year 2016
- csv: comma-separated values

Let's start exploring!

In [47]:
import pandas as pd
time_use_data = pd.read_csv('gss_tu2016_main_file.csv')
time_use_data.head()

CASEID  pumfid   wght_per  survmnth    wtbs_001  agecxryg  agegr10  \
0   10000   10000   616.6740         7    305.1159        96        5   
1   10001   10001  8516.6140         7      0.0000         6        5   
2   10002   10002   371.7520         1    362.7057         2        4   
3   10003   10003  1019.3135         3      0.0000        96        6   
4   10004   10004  1916.0708         9  11388.9706        96        2   

   agehsdyc  ageprgrd  chh0014c  ...  ree_02  ree_03  rlr_110  lan_01  \
0        62        96         0  ...       1       1        1       1   
1        32         5         0  ...       5       6        3       1   
2         9        10         3  ...       5       1        1       1   
3        65        96         0  ...       3       2        2       1   
4        25        96         0  ...       9      99        9       9   

   lanhome  lanhmult  lanmt  lanmtmul  incg1  hhincg1  
0        1         1      1         1      1        1  
1        5         2      5         2      5        8  
2        1         1      1         1      3        8  
3        1         1      1         1      2        2  
4       99         9     99         9      2        4  

[5 rows x 350 columns]

Where the heck did the nice names go?

Let's try to look at all of the column names:

In [48]:
time_use_data.columns

Index(['CASEID', 'pumfid', 'wght_per', 'survmnth', 'wtbs_001', 'agecxryg',
       'agegr10', 'agehsdyc', 'ageprgrd', 'chh0014c',
       ...
       'ree_02', 'ree_03', 'rlr_110', 'lan_01', 'lanhome', 'lanhmult', 'lanmt',
       'lanmtmul', 'incg1', 'hhincg1'],
      dtype='object', length=350)

Yikes. 350?

Let's make a smaller `DataFrame` with only the columns we care about. Here they are:

In order to create an easy to use dataset we will only keep the following columns:

- `CASEID` (participant ID), 
- `luc_rst` (large urban centre vs rural and small towns), 
- `agegr10` (age group), 
- `marstat` (marital status), 
- `sex` (sex), 
- `chh0014c` (number of kids 14 or under), 
- `gtu_110` (feeling rushed), and 
- `dur01` (duration spent sleeping). 

In [49]:
time_use_data['CASEID'].head()

0    10000
1    10001
2    10002
3    10003
4    10004
Name: CASEID, dtype: int64

Let's put all these column names in a list we'll name `important_columns`.

In [50]:
important_columns = ["CASEID", "luc_rst", "agegr10", "marstat", "sex",
                     "chh0014c", "gtu_110", "dur01"]

Just like we can extract a single column to get a `Series` , we can extract a subset of the columns to get a new `DataFrame`:

In [51]:
subset_time_use_data = time_use_data[important_columns]
subset_time_use_data.head()

CASEID  luc_rst  agegr10  marstat  sex  chh0014c  gtu_110  dur01
0   10000        1        5        5    1         0        1    510
1   10001        1        5        1    1         0        3    420
2   10002        1        4        1    2         3        1    570
3   10003        1        6        5    2         0        2    510
4   10004        1        2        6    1         0        1    525

### Still yucky to read

Here's a Python dictionary mapping icky names to better names:

In [52]:
column_names = {
    'CASEID': 'Participant ID',
    'luc_rst': 'Urban/Rural',
    'agegr10': 'Age Group',
    'marstat': 'Marital Status',
    'sex': 'Sex',
    'chh0014c': 'Kids under 14',
    'gtu_110': 'Feeling Rushed',
    'dur01': 'Sleep Duration'
}

In [53]:
column_names['marstat']

'Marital Status'

### Renaming columns!

This code makes a new `DataFrame` with the column names replaced (The `columns=` part is required magic):

In [54]:
time_use_subset = subset_time_use_data.rename(columns=column_names)
time_use_subset.head()

Participant ID  Urban/Rural  Age Group  Marital Status  Sex  Kids under 14  \
0           10000            1          5               5    1              0   
1           10001            1          5               1    1              0   
2           10002            1          4               1    2              3   
3           10003            1          6               5    2              0   
4           10004            1          2               6    1              0   

   Feeling Rushed  Sleep Duration  
0               1             510  
1               3             420  
2               1             570  
3               2             510  
4               1             525

# Subsetting rows

Two ways to subset by row.
+ By row index
+ By describing the columns to look in and the values to select for

## By row index: slicing

`DataFrame`s contain a variable called `iloc` that takes the row range and column range:

In [55]:
time_use_subset.iloc[0:3, 1:4] # Rows 0, 1, and 2; columns 1, 2, and 3

Urban/Rural  Age Group  Marital Status
0            1          5               5
1            1          5               1
2            1          4               1

As with list slicing, the second index is not included in the result.

## By using the `DataFrame` row labels

`DataFrame`s contain a variable called `loc` that uses the labels, not the indexes. Watch that second index!

(The row labels are integers in `time_use_subset`.)

In [56]:
time_use_subset.loc[0:3, 'Urban/Rural':'Marital Status']

Urban/Rural  Age Group  Marital Status
0            1          5               5
1            1          5               1
2            1          4               1
3            1          6               5

### `iloc` vs `loc`

+ With `iloc`, the second part of a range is not included.
+ With `loc`, the second part of a range __is__ included.

# In which rows is the Marital Status 1?

This gets a `Series` of Boolean values. Because `=` is used for assignment, Python uses `==` for equality.

In [57]:
time_use_subset['Marital Status'] == 1

0        False
1         True
2         True
3        False
4        False
         ...  
17385    False
17386    False
17387    False
17388    False
17389     True
Name: Marital Status, Length: 17390, dtype: bool

Ooh! That's fun. We can give this to `loc` as the row specification.

# Practice

A reminder of `time_use_subset` info:

In [58]:
time_use_subset.head()

Participant ID  Urban/Rural  Age Group  Marital Status  Sex  Kids under 14  \
0           10000            1          5               5    1              0   
1           10001            1          5               1    1              0   
2           10002            1          4               1    2              3   
3           10003            1          6               5    2              0   
4           10004            1          2               6    1              0   

   Feeling Rushed  Sleep Duration  
0               1             510  
1               3             420  
2               1             570  
3               2             510  
4               1             525

How do you get a `DataFrame` for rows where `Sex == 1`?

In [59]:
# Get the True/False Series from the column.
sex_1_series = time_use_subset['Sex'] == 1
sex_1_df = time_use_subset.loc[sex_1_series, :]
sex_1_df.head()

Participant ID  Urban/Rural  Age Group  Marital Status  Sex  Kids under 14  \
0           10000            1          5               5    1              0   
1           10001            1          5               1    1              0   
4           10004            1          2               6    1              0   
5           10005            1          1               6    1              2   
6           10006            1          1               6    1              1   

   Feeling Rushed  Sleep Duration  
0               1             510  
1               3             420  
4               1             525  
5               1             435  
6               4             635

How do you get a `DataFrame` for rows where `Sex == 1`, but with only the `Age Group`, `Marital Status`, and `Sex` columns?

In [60]:
sex_1_df = time_use_subset.loc[sex_1_series, 'Age Group':'Sex']
sex_1_df.head()

Age Group  Marital Status  Sex
0          5               5    1
1          5               1    1
4          2               6    1
5          1               6    1
6          1               6    1

How do you get a `DataFrame` with only `Feeling Rushed` and `Sleep duration` columns for everyone except people who have a `Feeling Rushed` values of `1`. (Hint: `!=` means "not equal to", the opposite of `==`.)

In [64]:
rushed_series = time_use_subset['Feeling Rushed'] != 1 # Get our column of True/False values
rushed_df = time_use_subset.loc[rushed_series, 'Feeling Rushed':'Sleep Duration']
rushed_df.head()

Feeling Rushed  Sleep Duration
1               3             420
3               2             510
6               4             635
7               5             440
8               2             525